src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -202,19 +275,350 @@ protected override ODataRequestMessageWrapper CreateRequestMessage(string method
+
+                if (index < descriptors.Count)
+                {
+                    lastVisitedDescriptor = descriptors[index];
+                }
+            }
+           
+            return lastVisitedDescriptor;
+        }
+
+        /// <summary>
+        /// Checks whether an entry has a data modification exception instance annotation.
+        /// </summary>
+        /// <param name="materializer">The materializer entry to check for the data modification exception instance annotation.</param>
+        /// <returns>true if the entry has the data modification exception, otherwise false.</returns>
+        private static bool HasFailedOperation(IMaterializerState materializer)
+        {
+            if (materializer is MaterializerDeletedEntry deltaEntry)
+            {
+                return deltaEntry.DeletedResource.InstanceAnnotations.Any(a => a.Name == XmlConstants.DataModificationException);
+            }
+            else if (materializer is MaterializerEntry entry)
+            {
+                return entry.Entry.InstanceAnnotations.Any(a => a.Name == XmlConstants.DataModificationException);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// This method creates an operation response for each operation.
+        /// </summary>
+        /// <param name="materializerState">An instance of the <see cref="IMaterializerState"/> entry associated with an operation response.</param>
+        /// <param name="isTopLevelResponse">true for a top-level response otherwise false.</param>
+        /// <exception cref="DataServiceRequestException"></exception>
+        private void CreateOperationResponse(IMaterializerState materializerState, bool isTopLevelResponse)
+        {
+            Exception ex = null;
+            OperationResponse operationResponse = null;
+            DataServiceResponse dataServiceResponse = new DataServiceResponse(
+                headers: null, 
+                statusCode: -1, 
+                this.operationResponses, 
+                batchResponse: false);
+            Descriptor descriptor = null;
+
+            try
+            {
+                if (this.adjacentDescriptors.Count == 0)
+                {
+                    descriptor = this.bulkUpdateGraph.TopLevelDescriptors[0];
+                    this.adjacentDescriptors.Push((0, this.bulkUpdateGraph.TopLevelDescriptors));
+                }
+                else
+                {
+                    Descriptor lastVisitedDescriptor = this.GetLastVisitedDescriptor();
+
+                    if (lastVisitedDescriptor != null)
+                    {
+                        IReadOnlyList<Descriptor> relatedDescriptors = this.bulkUpdateGraph.GetRelatedDescriptors(lastVisitedDescriptor);
+
+                        if (relatedDescriptors.Count == 0)
+                        {
+                            if (this.adjacentDescriptors.Peek().Item1 == this.adjacentDescriptors.Peek().Item2.Count - 1)

```

Comment:
If we had aliases for the tuple fields, we could have used those here instead of `Item1` and `Item2`.
Also, these manipulations of the `adjacentDecriptors` stack are not obvious. I think you should decorate them with commentary that explains the intend and what's going on. Or you create wrapper helper methods that encapsulate the logic. Or both.

Reply:
Like in this case, it's not obvious that we're removing the last level of descriptors from the stack because we have visited every descriptor at that level.

Reply:
+1 on adding comments to help someone understand what is happening